### PR TITLE
Avoid doc.write if a 3p frame has already been instrumented

### LINKED ISSUE #9249: Blank ads caused by doc.write in environment.js
`doc.write('<script>window.parent.ampManageWin(window)</script>');` is making ads blank.
(we might override doc.close too late?)

screen shot 2017-05-05 at 3 53 42 pm

### PR DESCRIPTION
**Pull request by lannka**
Fixes #9249

@cramforce , do you remember why we had the `instrumentDocWrite` code? 

`doc.write('<script>window.parent.ampManageWin(window)</script>');` is making ads blank.
(we might override doc.close too late?)

screen shot 2017-05-05 at 3 53 42 pm

### CODE REVIEW

**Comment by jridgewell** [comment]
Bad ads installing video stream decoding JS.

**Comment by lannka** [comment]
@jridgewell why [manageWin](https://github.com/ampproject/amphtml/blob/master/3p/environment.js#L150) is not enough and we have to [instrumentDocWrite](https://github.com/ampproject/amphtml/blob/master/3p/environment.js#L148)?

**Comment by jridgewell** [comment]
http://stackoverflow.com/questions/3216871/why-does-using-document-write-to-create-an-iframe-cancel-out-all-remaining-jav

**Comment by lannka** [comment]
@jridgewell don't quite understand your point.

**Comment by jridgewell** [comment]
It's another defense against badly written code.

**Comment by lannka** [comment]
the mystery to me is that we already did `manageWin(iframe.contentWindow)` immediately when we saw a new iframe window. but for some reasons, we did the same thing again and again at different time:
1. `iframe.onload = function () { /* one time here */}`
2. `iframe.contentWindow.document.close = function () { /* one more time here */}`
 
In my local test, I found 1) is necessary, though I don't have explanation. In fact, when I try to debug, it becomes unnecessary.
2) seems to be not needed, or maybe I wasn't using an appropriate test case.

So really want to hear from @cramforce who introduced those hacks :-). What problems were there?

File: 3p/environment.js [hunk 1]
```diff
@@ -178,13 +142,15 @@ function instrumentEntryPoints(win) {
   // Change setTimeout to respect a minimum timeout.
   const setTimeout = win.setTimeout;
   win.setTimeout = function(fn, time) {
+    dev().info('SETTIMEOUT', win.name);
```
Comment (dvoytenko) [inline 3p/environment.js]: `DO NOT SUBMIT`?

**Comment by dvoytenko** [comment]
@lannka Do we know the reason why it breaks anything, however?

**Comment by cramforce** [comment]
Are we running the code twice on the same window. There is a case of removing all of this code now that browsers throttle themselves more aggressively.

**Comment by lannka** [comment]
>@lannka Do we know the reason why it breaks anything, however?

@dvoytenko as I mentioned in PR description
> doc.write('<script>window.parent.ampManageWin(window)</script>'); is making ads blank.
(we might override doc.close too late?)

See that screenshot.

> Are we running the code twice on the same window. There is a case of removing all of this code now that browsers throttle themselves more aggressively.

@cramforce No. `manageWin()` is flag protected (`window.ampSeen`). However, the `doc.write` is not. I can do
```js
if (!parent.ampSeen) {
  doc.write('<script>window.parent.ampManageWin(window)</script>');		
}
```
But I want to understand what corner case this code is trying to handle, can we get rid of them completely? At least the `doc.write` part, which seems dangerous.

**Comment by cramforce** [comment]
That conditional should definitely be added!!

There are many ways how iframes are added to ad pages. This is handling one of those cases. Might or might not be needed for doubleclick.

**Comment by lannka** [comment]
@cramforce adding the check fixes the issue. this is good for now, let's follow up offline see if we can improve some of the code here.

**Comment by lannka** [comment]
updated description

**Pull request merged**